UPGRADE.md

```diff
@@ -0,0 +1,118 @@
+If you generate code that should run in PHP 5.x, it is advisable to strip
+`string`, `int`, `float` and `bool` from type definitions passed to
+`Zend\Code\ParameterGenerator` instances.
+
+### `Zend\Code\Reflection\ParameterReflection#getType()` changes
+
+PHP 7 introduced [`ReflectionParameter#getType()`](http://php.net/manual/en/reflectionparameter.gettype.php).
+
+In order to not override this method, `Zend\Code\Reflection\ParameterReflection#getType()`
+was renamed to `Zend\Code\Reflection\ParameterReflection#detectType()`.
+
+If you relied on `Zend\Code\Reflection\ParameterReflection#getType()`, you can
+simply replace the method calls in your code.
+ 
+### DocBlock types ignored by `Zend\Code\Generator\ParameterGenerator::fromReflection()`
+
+As a direct consequence of the previous change, calls to 
+`Zend\Code\Generator\ParameterGenerator::fromReflection()` will not mirror the
+type hints read from a method's DocBlock.
+
+As an example, take following code:
+
+```php
+class Foo
+{
+    /**
+     * @param string $baz
+     */
+    public function bar($baz)
+    {
+    }
+}
+
+$methodGenerator = \Zend\Code\Generator\MethodGenerator::fromReflection(
+    new \Zend\Code\Reflection\MethodReflection('Foo', 'bar')
+);
+
+var_dump($methodGenerator->getParameters()[0]->getType());
+```
+
+In version 2.x, this code produces `"string"`, in version 3.x it returns `null`.
+
+### Type strings are validated
+
+If you attempt to generate type-hints for parameters or return types, those types are
+now validated before the code is generated.
+
+Be sure to check which values you pass to `Zend\Code\Generator\MethodGenerator#setReturnType()`
+or `Zend\Code\Generator\ParameterGenerator#setType()`, as you may incur in a
+`Zend\Code\Generator\Exception\InvalidArgumentException` being thrown if any
+of those types are invalid strings:
+
+```php
+$parameterGenerator->setType('foo'); // valid
+$parameterGenerator->setType('array'); // valid
+$parameterGenerator->setType('bool'); // valid
+$parameterGenerator->setType('123'); // invalid (throws exception)
+$parameterGenerator->setType(''); // invalid (throws exception)
+$parameterGenerator->setType('*'); // invalid (throws exception)
+$parameterGenerator->setType('\\'); // invalid (throws exception)
+```
+
+
+### Generated type-hints are now prefixed by `"\"`
+
+Generated type-hints are now prefixed with the `NAMESPACE_SEPARATOR`,
+`"\"`.
+
+Take following example code:
+
+```php
+$parameter = new \Zend\Code\Generator\ParameterGenerator('bar', 'baz');
+$method    = new \Zend\Code\Generator\MethodGenerator('foo', [$parameter]);
+
+$method->setReturnType('tab');
+
+echo $method->generate();
+```
+
+This code produces `public function foo(baz $bar) {}` in 2.x.
+In version 3.x, it produces `public function foo(\baz $bar) : \tab {}`.
+
+In order to avoid migration problems, be sure to always pass fully qualified class
+names to `Zend\Code\Generator\MethodGenerator` and `Zend\Code\Generator\ParameterGenerator`.
```

Comment:
Another question: if we're generating the method, presumably it's within a class, correct? Why are we not doing import statements and using resolved names? (I suspect it may be to prevent conflicts, but I'm curious what your rationale was for this change.)